content/post/2022-04-25-PHP-database-result-iteration.md

```diff
@@ -0,0 +1,164 @@
+---
+title: Best practices in PHP when iterating over database results
+date: 2022-04-25
+tags:
+  - PHP
+  - database
+  - history
+  - patterns
+description: "How the default method for iterating over database results
+is stuck in 2005, goes against best practices and what to do about it."
+---
+
+Many PHP database libraries use a `while` loop with assignment to iterate
+over database results. In this article I'll show why this pattern is so
+pervasive, why it opens the door for errors and what
+you can use instead.
+
+<!--more-->
+
+This is a typical piece of PHP code, using [PDO](https://www.php.net/manual/en/book.pdo.php) for accessing the database and iterating over the result:
+
+```php
+$result = $db->query("SELECT id, name from my_table");
+while ($row = $result->fetch()) {
+    // do something with $row
+}
+```
+
+Many coding styles don't recommend the practice of "assignment inside control structure epxresssion". The [MediaWiki Coding Style](https://www.mediawiki.org/wiki/Manual:Coding_conventions/PHP) even has [a rule that forbids assignments in control structures](https://github.com/wikimedia/mediawiki-tools-codesniffer/blob/master/MediaWiki/Sniffs/ControlStructures/AssignmentInControlStructuresSniff.php), which prevents this style of looping. Which could raise the questions:
+
+- Why disallow assignments in control structures?
+- How do I fix the code to make the coding style checker happy?
+- How did this style of looping evolve and why did the evolution stop?
+
+### Why assignments in control structures are dangerous
+The danger comes from the "overloading" of the equal sign in an expression:
```

Comment:
I'm afraid I don't understand why this talks about "overloading". Sure, some languages like BASIC use the same symbol `=` for both assignments and comparisons. But PHP doesn't. It's not "overloaded".